File: Colorization/colorizeVideo.py
#!/usr/bin/env python3
"""Colorize a video with OpenCV DNN and an ONNX model."""

from __future__ import annotations

import argparse
import time
from pathlib import Path

import cv2 as cv

from colorization import DEFAULT_MODEL, colorize_frame, load_network, validate_output


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(
        description="Colorize a video frame by frame with OpenCV DNN."
    )
    parser.add_argument(
        "--input",
        type=Path,
        default=Path(__file__).resolve().parent / "greyscaleVideo.mp4",
        help="Input video (default: the bundled sample).",
    )
    parser.add_argument(
        "--model",
        type=Path,
        default=DEFAULT_MODEL,
        help="Path to colorization_eccv16.onnx.",
    )
    parser.add_argument(
        "--output",
        type=Path,
        default=Path("colorized-video.avi"),
        help="Destination video.",
    )
    parser.add_argument(
        "--max-frames",
        type=int,
        default=0,
        help="Stop after this many frames; 0 processes the entire video.",
    )
    parser.add_argument(
        "--no-display",
        action="store_true",
        help="Do not open an OpenCV preview window.",
    )
    parser.add_argument(
        "--validate",
        action="store_true",
        help="Validate each generated frame and the final frame count.",
    )
    return parser.parse_args()


def main() -> int:
    args = parse_args()
    capture = cv.VideoCapture(str(args.input))
    if not capture.isOpened():
        raise FileNotFoundError(f"Could not open input video: {args.input}")

    width = int(capture.get(cv.CAP_PROP_FRAME_WIDTH))
    height = int(capture.get(cv.CAP_PROP_FRAME_HEIGHT))
    fps = capture.get(cv.CAP_PROP_FPS)
    if not fps or not 1.0 <= fps <= 240.0:
        fps = 30.0

    args.output.parent.mkdir(parents=True, exist_ok=True)
    writer = cv.VideoWriter(
        str(args.output),
        cv.VideoWriter_fourcc(*"MJPG"),
        fps,
        (width, height),
    )
    if not writer.isOpened():
        capture.release()
        raise RuntimeError(f"Could not open output video: {args.output}")

    network = load_network(args.model)
    processed = 0
    inference_seconds = 0.0
    try:
        while True:
            ok, frame = capture.read()
            if not ok:
                break

            start = time.perf_counter()
            output, chroma_score = colorize_frame(frame, network)
            inference_seconds += time.perf_counter() - start
            if args.validate:
                validate_output(frame, output, chroma_score)

            writer.write(output)
            processed += 1
            if not args.no_display:
                cv.imshow("Colorized video", output)
                if cv.waitKey(1) & 0xFF == 27:
                    break
            if args.max_frames and processed >= args.max_frames:
                break
    finally:
        capture.release()
        writer.release()
        if not args.no_display:
            cv.destroyAllWindows()

    if args.validate and processed == 0:
        raise RuntimeError("No frames were processed.")

    average = inference_seconds / processed if processed else 0.0
    print(f"Saved {processed} frames to {args.output}")
    print(f"Average inference time: {average:.3f} seconds per frame")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: Colorization/colorizeVideo.cpp
// This code is written by Sunita Nayak at BigVision LLC.
// It is subject to the license terms in the LICENSE file in this folder.

#include "colorization.hpp"

#include <opencv2/highgui.hpp>
#include <opencv2/videoio.hpp>

#include <chrono>
#include <filesystem>
#include <iomanip>
#include <iostream>
#include <stdexcept>

int main(int argc, char** argv) {
    try {
        const colorization::CommonOptions options = colorization::parseOptions(
            argc,
            argv,
            "greyscaleVideo.mp4",
            "colorized-video.avi",
            true);

        cv::VideoCapture capture(options.input.string());
        if (!capture.isOpened()) {
            throw std::runtime_error(
                "Could not open input video: " + options.input.string());
        }
        const int width =
            static_cast<int>(capture.get(cv::CAP_PROP_FRAME_WIDTH));
        const int height =
            static_cast<int>(capture.get(cv::CAP_PROP_FRAME_HEIGHT));
        double fps = capture.get(cv::CAP_PROP_FPS);
        if (!std::isfinite(fps) || fps < 1.0 || fps > 240.0) {
            fps = 30.0;
        }

        const std::filesystem::path parent = options.output.parent_path();
        if (!parent.empty()) {
            std::filesystem::create_directories(parent);
        }
        cv::VideoWriter writer(
            options.output.string(),
            cv::VideoWriter::fourcc('M', 'J', 'P', 'G'),
            fps,
            cv::Size(width, height));
        if (!writer.isOpened()) {
            throw std::runtime_error(
                "Could not open output video: " + options.output.string());
        }

        cv::dnn::Net network = colorization::loadNetwork(options.model);
        int processed = 0;
        double inferenceSeconds = 0.0;
        cv::Mat frame;
        while (capture.read(frame)) {
            const auto start = std::chrono::steady_clock::now();
            auto [output, chromaScore] =
                colorization::colorizeFrame(frame, network);
            const std::chrono::duration<double> elapsed =
                std::chrono::steady_clock::now() - start;
            inferenceSeconds += elapsed.count();
            if (options.validate) {
                colorization::validateOutput(frame, output, chromaScore);
            }

            writer.write(output);
            ++processed;
            if (!options.noDisplay) {
                cv::imshow("Colorized video", output);
                if ((cv::waitKey(1) & 0xFF) == 27) {
                    break;
                }
            }
            if (options.maxFrames > 0 && processed >= options.maxFrames) {
                break;
            }
        }

        capture.release();
        writer.release();
        if (!options.noDisplay) {
            cv::destroyAllWindows();
        }
        if (options.validate && processed == 0) {
            throw std::runtime_error("No frames were processed.");
        }

        const double average =
            processed > 0 ? inferenceSeconds / processed : 0.0;
        std::cout << "Saved " << processed << " frames to " << options.output
                  << '\n'
                  << "Average inference time: " << std::fixed
                  << std::setprecision(3) << average
                  << " seconds per frame\n";
        return 0;
    } catch (const std::exception& error) {
        std::cerr << "Error: " << error.what() << '\n';
        return 1;
    }
}
